### ISSUE TITLE
Exoscale provisioning

### ISSUE THREAD

**Issue by fretscha** [rancher/rancher#13291]
Created node template and entered Exoscale API key and secret. When I try to create a cluster on Exoscale, based on this template, I get following error:

> Error setting machine configuration from flags provided: missing an API key (--exoscale-api-key) or API secret key (--exoscale-api-secret-key); Timeout waiting for ssh key

Unsucessfully Tried multiple keypairs & multiple accounts. 

Docker-machine provisioning from command line works. Rancher UI fails.
 
---
| Useful | Info |
| :-- | :-- |
|Versions|Rancher `v2.0.0` UI: `v2.0.41` |
|Access|`local` `admin`|
|Route|`authenticated.cluster.nodes.index`|

**Comment by fretscha** [comment]
Problem still exists in v2.0.2

**Comment by justjulian** [comment]
Same here with v2.0.6

**Comment by greut** [comment]
Oh crickey. We made changes in the 1.6 branch which aren't as of yet been ported to the 2.x world. We'll take care of that soonish.

**Comment by greut** [comment]
As the "node provider" still uses docker-machine which is in a bugfix mode for a long time already. It's probably easier to rely on the _bring your own nodes_ cluster setup.

![screenshot from 2018-07-31 09-06-12](https://user-images.githubusercontent.com/1388/43443407-0e71bc12-94a1-11e8-8484-679cb4fff9f6.png)

A working cloud-init could be the following one.

```yaml
#cloud-config

manage_etc_hosts: true

apt_sources:
- source: "deb [arch=amd64] https://download.docker.com/linux/ubuntu xenial stable"
  keyid: 9DC858229FC7DD38854AE2D88D81803C0EBFCD88

package_update: true
package_upgrade: true

packages:
- [ docker-ce, "17.03.2~ce-0~ubuntu-xenial" ]


runcmd:
- [ usermod, -aG, docker, ubuntu ]
```

NB: we are working to bring back the node provider as well.

**Comment by Remigius2011** [comment]
(same here)

When will the changes be available in Rancher - is the node driver dynamically loaded from somewhere or do we have to wait for a new Rancher version?
Will it be possible to select a zone/location (here, it selected "CH-DK-2" without my intervention - I'd like to control where my data stays - otherwise I can as well use any infrastructure provider).
How can I use the so-called "cloud-init" given above - if possible with a given zone?
Is terraform currently working? Then I could at least try to use iac to provision my environment,

**Comment by greut** [comment]
@Remigius2011 some changes have been merged to the ui, https://github.com/rancher/ui/pull/2132 which enables the zone selection.

As for the terraform provider, it should work. https://github.com/exoscale/terraform-provider-exoscale/tree/master/examples/rke

We are in the process to get the provider moved upstream but things may take some time as hashicorp is problably focused on their 0.12 release. Cheers,

**Comment by greut** [comment]
@fretscha do you still have the issue using rancher/rancher:v2.0.7-rc2 ?

**Comment by Remigius2011** [comment]
Is this supposed to be fixed in 2.0.7?

**Comment by fretscha** [comment]
@greut  The issue of the box setup looks like to be resolved in rancher/rancher:v2.0.7-rc2.
Some firewall rules (sg) are still not complete / missing.

**Comment by Remigius2011** [comment]
@fretscha I have just retestet with rancher/rancher:v2.0.8. Unfortunately, exoscale provisioning still does not work for me. I think I am not doing anything exotic - just a single node, medium, 50G, ch-dk-2. The log is here (remark1: this is just one iteration - it retries to provision in a loop which provides the same entries except for the time stamps, remark2: I have replaced potentially sensitive data like UIDs, IP addresses etc. with start, remark3: sorry, it looks like the intended formatting is broken by github):

> 2018/08/27 08:41:18 [INFO] stdout: Creating CA: management-state/node/nodes/wrk-1/certs/ca.pem
> 2018/08/27 08:41:18 [INFO] stdout: Creating client certificate: management-state/node/nodes/wrk-1/certs/cert.pem
> 2018/08/27 08:41:19 [INFO] stdout: Running pre-create checks...
> 2018/08/27 08:41:19 [INFO] stdout: Creating machine...
> 2018/08/27 08:41:19 [INFO] stdout: (wrk-1) Querying exoscale for the requested parameters...
> 2018/08/27 08:41:20 [INFO] stdout: (wrk-1) Generate an SSH keypair...
> 2018/08/27 08:41:20 [INFO] stdout: (wrk-1) Spawn exoscale host...
> 2018/08/27 08:41:20 [INFO] stdout: (wrk-1) Deploy &egoscale.DeployVirtualMachine{Account:"", AffinityGroupIDs:[]string{}, AffinityGroupNames:[]string(nil), CustomID:"", DeploymentPlanner:"", Details:map[string]string(nil), DiskOfferingID:"", DisplayName:"wrk-1", DisplayVM:(*bool)(nil), DomainID:"", Group:"", HostID:"", Hypervisor:"", IP4:(*bool)(nil), IP6:(*bool)(nil), IP6Address:net.IP(nil), IPAddress:net.IP(nil), IPToNetworkList:[]egoscale.IPToNetwork(nil), Keyboard:"", KeyPair:"docker-machine-wrk-1", Name:"wrk-1", NetworkIDs:[]string(nil), ProjectID:"", RootDiskSize:50, SecurityGroupIDs:[]string{"******}, SecurityGroupNames:[]string(nil), ServiceOfferingID:"******", Size:0, StartVM:(*bool)(nil), TemplateID:"******", UserData:"******", ZoneID:"******"}
> 2018/08/27 08:41:41 [INFO] stdout: (wrk-1) IP Address: ***.***.***.***, SSH User: ******
> 2018/08/27 08:41:44 [INFO] stdout: Waiting for machine to be running, this may take a few minutes...
> 2018/08/27 08:41:44 [INFO] stdout: Detecting operating system of created instance...
> 2018/08/27 08:41:44 [INFO] stdout: Waiting for SSH to be available...
> 2018/08/27 08:41:45 [INFO] stdout: Detecting the provisioner...
> 2018/08/27 08:41:45 [INFO] stdout: Provisioning with ubuntu(systemd)...
> 2018/08/27 08:41:56 [INFO] stdout: Installing Docker...
> 2018/08/27 08:42:11 [INFO] stdout: Copying certs to the local machine directory...
> 2018/08/27 08:42:13 [INFO] stdout: Copying certs to the remote machine...
> 2018/08/27 08:42:14 [INFO] stdout: Setting Docker configuration on the remote daemon...
> 2018/08/27 08:42:17 [INFO] stdout: Checking connection to Docker...
> 2018/08/27 08:42:40 [INFO] stdout: The default lines below are for a sh/bash shell, you can specify the shell you're using, with the --shell flag.
> 2018/08/27 08:42:40 [INFO] stdout:
> 2018/08/27 08:42:40 [INFO] Generating and uploading node config wrk-1
> 2018/08/27 08:42:40 [ERROR] NodeController c-*****/m-***** [node-controller] failed with : Error creating machine: Error checking the host: Error checking and/or regenerating the certs: There was an error validating certificates for host "***.***.***.***:****": dial tcp ***.***.***.***:****: i/o timeout
> 2018/08/27 08:42:40 [INFO] NodeController c-*****/m-***** completed with dropped err: [node-controller] failed with : Error
> 2018/08/27 08:42:41 [INFO] NodeController c-*****/m-***** completed with dropped err: [node-controller] failed with : Error

**Comment by greut** [comment]
@fretscha we will take a look about those missing rules.

@Remigius2011 where do you run rancher from? The nodes have to must be able to talk to it. It's a :chicken: and :egg: problem where you need a public instance to spawn more instances.

I have an instance running rancher 2.0, that I access to via its IP address and <http://xip.io/>. Then that instance spawned the following cluster.

![screenshot from 2018-08-27 11-32-59](https://user-images.githubusercontent.com/1388/44652728-fd30a600-a9ec-11e8-9259-146c65a98a61.png)

![screenshot from 2018-08-27 11-34-26](https://user-images.githubusercontent.com/1388/44652814-38cb7000-a9ed-11e8-964f-f44316ebb237.png)

![screenshot from 2018-08-27 11-32-23](https://user-images.githubusercontent.com/1388/44652729-fd30a600-a9ec-11e8-918d-aeb6fcf1103f.png)

**Comment by Remigius2011** [comment]
@greut my rancher runs on a machine that is accessible via an nginx reverse proxy - configured according to rancher recommendations - listening on a public IP having a public DNS name. I am able to provision clusters on digitalocean with this setup. If necessary, I can provide more information over a private channel.

**Comment by greut** [comment]
@Remigius2011 support@exoscale.ch would be the perfect channel for this. Thanks!

**Comment by Remigius2011** [comment]
@greut Thanks for your fast and accurate support via support@exoscale.ch. As you mentioned in the answer to my support call, the reason was a pre-existing security group `rancher-machine` which did not contain the correct rules. After deleting it, I could successfully in create a cluster (just a single node for testing) using the exoscale provider.

One minor remark: in the list of node templates the location is displayed as "Unknown", as follows:

![exo-node-templates](https://user-images.githubusercontent.com/2393965/44666872-a8efeb00-aa19-11e8-9e77-026c5d9c6025.png)

Obviously, it would be nice to see the chosen location.

**Comment by pierre-emmanuelJ** [comment]
there is some port change
https://github.com/rancher/ui/pull/2194

**Comment by loganhz** [comment]
Please let us know in a comment if you still can not add a cluster with Exoscale node template , and we'll reopen it

For the the location is displayed, please submit a new issue.

Thanks a lot!

**Comment by Remigius2011** [comment]
Hi, same problem again with rancher 2.1.1. I was unable to delete the `rancher-machine` security group in the exoscale gui. Specifying another one to be created by the node driver (I would not know the rules to create one from scratch...) is also not supported.

**Comment by Remigius2011** [comment]
Meanwhile I was able to regenerate the `rancher-machine` security group, but this did not fix the problem this time. I am always stuck in a loop while provisioning the cluster.

**Comment by Remigius2011** [comment]
see also my exoscale Support Ticket #1023093.

**Comment by Remigius2011** [comment]
As mentioned in my last message before the issue was reopened (and in the support ticket), I managed to regenerate the `rancher-machine` security group.

The Exoscale node driver works in 2.0.8, but ceased again to work in 2.1.1 (I didn't check 2.1.0 as this was never a stable version).

Please check also whether the code that fixed the problem was cross-merged to the 2.1.x, 2.2.x and main branches (and whatever branches are necessary to make the fix available in future releases).

**Comment by Remigius2011** [comment]
It looks like it is not really dependent on the rancher version. I have now a new rancher v 2.0.8 cluster running which fails with the same pattern, whereas in the other - also v 2.0.8 - it works.

**Comment by Remigius2011** [comment]
I'm about to check differences in network configuration etc. - I'll keep you posted on the result.

**Comment by loganhz** [comment]
I can add a cluster with exoscale node template successfully in v2.1.1

I guess if the user who upgrades Rancher from v2.0.X might have the `rancher-machine` security group issue.

Workaround is 

**Option 1**: Please see https://github.com/rancher/rancher/issues/13291#issuecomment-431868108 

**Option 2**: Delete `rancher-machine` in exoscale console and create a node template in UI again. It will send out request to create a new `rancher-machine` group with proper rules

**Options 3**: Config your own security group in  exoscale console with the rules https://user-images.githubusercontent.com/15922119/44715367-ec993200-aab7-11e8-83f1-5c7dfefabbb9.png and select it as the security group in Rancher UI

**Comment by Remigius2011** [comment]
@loganhz thanks for your suggestions.

Since yesterday evening (see my [comment above](https://github.com/rancher/rancher/issues/13291#issuecomment-431895357)) I cannot exclude it is a network problem, as it also occurs on a v2.0.8 rancher that I have installed on the same machine. I have already tried your suggestions 1 and 2. I have also meanwhile checked the following configurations on both rancher nodes without noticing a difference that could lead to the observed behavior, i.e. one node can create clusters and the other fails (as a side note that may make everybody go `ahhh!` I must tell both rancher nodes run behind an `nginx` reverse proxy):

* `/etc/network/interfaces` on the rancher nodes
* proxy environment variables on nodes (identical) and inside rancher containers (both none)
* nginx site configuration (identical except for the host names and the certificates)
* host certificates (identical except for the key pair)
* rest of nginx config (no asymmetries between the nodes found)
* docker command line to start rancher (identical)
* firewall configuration (both IP addresses belong to the same alias and are thus treated equal)

Also, as I have discovered this morning, the problem is not restricted to Exoscale, also the creation of a digitalocean cluster fails on the new rancher server (albeit with a different error message).

I will try now to set up a custom cluster and see whether I find a more detailed error message on the log output of the cluster node.

**Comment by Remigius2011** [comment]
the behavior with custom clusters is at least consistent with node-provider provisioned ones: it succeeds on the 'good' node and fails on the 'bad' one - with:

```
[INFO] cluster [c-6glbm] provisioning: [healthcheck] Start Healthcheck on service [kube-apiserver] on host [x.x.x.x]
```

then a little while later:

```
[ERROR] cluster [c-6glbm] provisioning: [controlPlane] Failed to bring up Control Plane: Failed to verify healthcheck: Failed to check https://localhost:6443/healthz for service...
```

at least this mostly excludes that it is the same problem as in the beginning of this issue - so I'll stop reporting and continue my experiments silently...

**Comment by loganhz** [comment]
Can you try to clean up the 'bad' node first and re-add it again, please?

You can find the clean up script via https://gist.github.com/superseb/2cf186726807a012af59a027cb41270d

**Comment by Remigius2011** [comment]
**tl;dr** - I think you can close this issue (as far as I am concerned).

The problem problem is solved: as I already suspected some time before, it was not the exoscale node driver, but last night discovered a problem in the `nginx` site configuration of the reverse proxy (aka external load balancer). I am adding the details here - even though it does not really fit into this issue - for other people arriving here via search engine who might have a similar problem.

As I had a running instance, I was assuming the new instance would work with the same `nginx` configuration, but the running one was stepwise upgraded from v2.0.2 to v2.0.8, so it had a different history than the new one and hence a different state.

In the server sections for http and https ingress traffic, I had - historically - listen statements not only for ipv4, but also for ipv6, e.g.:

```
  listen [::]:443 ssl http2;
```

As soon as I have removed them, it worked (and still works).

Also, as suggested [here](https://rancher.com/docs/rancher/v2.x/en/installation/single-node/single-node-install-external-lb/#2-choose-an-ssl-option-and-install-rancher) I added the parameter `--no-cacerts` to the `docker run` command that starts the rancher server (but I did not retest to verify whether this had any influence - initially it didn't but adding it will not hurt if you have an ssl certificate on the reverse proxy).

The correct configuration is described [here](https://rancher.com/docs/rancher/v2.x/en/installation/single-node/single-node-install-external-lb/#example-nginx-configuration) and it may make sense to check once in a while for changes and align even working configurations with this example.

***Thanks and sorry*** to everybody who has helped me to chase the phantom and solve this riddle.

**Comment by loganhz** [comment]
Thanks for the update! Close this issue
